replica/table.cc

```diff
@@ -2627,15 +2627,11 @@ future<row_locker::lock_holder> table::push_view_replica_updates(shared_ptr<db::
 
 future<row_locker::lock_holder> table::do_push_view_replica_updates(shared_ptr<db::view::view_update_generator> gen, schema_ptr s, mutation m, db::timeout_clock::time_point timeout, mutation_source source,
         tracing::trace_state_ptr tr_state, reader_concurrency_semaphore& sem, const io_priority_class& io_priority, query::partition_slice::option_set custom_opts) const {
-    if (!_config.view_update_concurrency_semaphore->current()) {
-        // We don't have resources to generate view updates for this write. If we reached this point, we failed to
-        // throttle the client. The memory queue is already full, waiting on the semaphore would cause this node to
-        // run out of memory, and generating hints would ultimately result in the disk queue being full too. We don't
-        // drop the base write, which could create inconsistencies between base replicas. So we dolefully continue,
-        // and note the fact we dropped a view update.
-        ++_config.cf_stats->dropped_view_updates;
-        co_return row_locker::lock_holder();
-    }
+    // Make sure that we have some resources available to generate the view update.
+    // If the database has been overwhelemed to the point where getting a few units timeouts,
+    // then it just can't handle the write at this moment. In such situation the query should timeout.
+    auto view_update_permit_units = co_await seastar::get_units(*_config.view_update_concurrency_semaphore, 1024, timeout);
```

Comment:
Just a timeout is not enough here. We also need a cap on how much memory the queue can consume. When accounting for this we must consider the queue in the semaphore itself, as well as the task objects that this waiting create, and finally the size of the mutation that is kept around.

When the queue size reaches the cap, we should start dropping view updates, just like before.

Reply:
Why do we have to drop view updates? Can't we wait until a place in the queue frees up?

How is the number of normal writes limited? Is the amount of mutations applied simlutanously limited to some fixed amount? The size of queue created by things that wait on the `view_update_concurrency_semaphore` is limited by the amount of mutations that are being applied at once, which should be already limited somehow I think.

This part of the code acquires just 1024 units, but later we acquire more + the size of the mutations.

Reply:
The number of writes we accept are limited indirectly by the rate at which we can flush these writes to disk. If we cannot keep up, the writes are blocked (queued). This queue doesn't have a hard upper size limit, but regular writes are synchronous: assuming a well behaving client with a limited concurrency, after all its write fibers are in wait state, there won't be any more writes coming until the replica can catch up and accept new writes.
Since view updates are asynchronous, this doesn't apply: if the paired view replica can't keep up, the view update queues can grow to infinity (OOM).

Reply:
> Hmm that's troublesome. The way I imagined overload control to work between two remote processes is that there's some sort of bounded size queue for requests. One process submits requests onto this queue and the other one consumes as many of them as it can process. If the queue fills up then the submitter has to wait. The queue could be just a TCP connection and to enact backpressure the consumer can stop reading from the connection for a moment.

What you describe here is the case of a well-behaving client which has limited concurrency. Unfortunately, materialized view updates are equivalent to a client with unlimited concurrency, even if the end client generating the base table writes is a well behaving one, with limited concurrency.
The only way to deal with such a client with unlimited concurrency is to engineer a pressure release valve into the system, which drops requests when the incoming rate of requests is larger than the consumer's rate of processing them. Throttling alone (regardless of which layer you implement it in) just doesn't work, because waiting requests pile up in the base replica's memory in an infinite queue, with no way to slow down incoming base table writes.

Reply:
> What you describe here is the case of a well-behaving client which has limited concurrency. Unfortunately, materialized view updates are equivalent to a client with unlimited concurrency, even if the end client generating the base table writes is a well behaving one, with limited concurrency.

Not really, even with unlimited concurrency the client probably sends the requests over a limited number of TCP connections. The CQL protocol handles 32k simultaneous requests on a single connection at once. If the client were to open an unlimited number of connections I guess we could stop listening for new ones.
The fact that the number of requests is unlimited doesn't mean that we can't stop reading them from the TCP connections. We can, and the client will have to wait before writing more data there.

Reply:
Btw I think that in a situation where there are so many base updates that we can't generate view updates in time (get the semaphore units), then it's more reasonable to fail the base update than it's to drop the view update. We're overloaded anyway, but at least we can try not to cause view inconsistencies.

If we got backpressure between base and view replicas right, the backpressure between the client and base replica could consist of waiting on the semaphore with a timeout and failing the base write if it can't be acquired in time.

Reply:
> Btw I think that in a situation where there are so many base updates that we can't generate view updates in time (get the semaphore units), then it's more reasonable to fail the base update than it's to drop the view update. We're overloaded anyway, but at least we can try not to cause view inconsistencies.

Which base update? Since base updates don't wait for view updates in any capacity, we cannot reliably fail the base update that generated the view update we are about to discard. So what base update do we fail? Do we fail a random one which happens to be executing at that time?

> If we got backpressure between base and view replicas right, the backpressure between the client and base replica could consist of waiting on the semaphore with a timeout and failing the base write if it can't be acquired in time.

This is basically a synchronous view, no?